### ISSUE TITLE
replace current job module with job submit service

### ISSUE THREAD

**Issue by grondo** [flux-framework/flux-core#338]
Job module on rank 0 should be replaced by a service implementing a suitable subset of Job Submit API (#268) and Job Feasibility Filter/Checks (#269).

The current `job` service should be extended to take a minimum viable job description encoded in JSON and perform a simple job feasibility check. If feasibility fails, a descriptive error is returned to the caller in the response. If job request is feasible, then the new job/submit module shall allocate a new lwj id and populate a job description skeleton in appropriate kvs directory, then respond to caller with success and the newly allocated jobid or kvs path.

The key requirements in this issue are the job feasibility check and response that indicates error (job not feasible) or success (a valid job request has been submitted to the queue)

OPTION: submit service has configurable or registered stack of filters that can reject and/or modify job submission.

**Comment by grondo** [comment]
Ok, I'd like to implement something quickly here that supports work for #295, and implements the service described above.

Going back to #268, we had proposed an off-the-cuff job submit protocol like:
```
topic: "job.request"
payload: {
  "cmdline"   : [ "program", "args", ],
  "cwd"       : "path",
  "environ"   : {  "var": "value", }, # (overrides or complete)
  "resources" : { request-obj },      # resource request object to be defined,
  "options"   : { option: "value" },  # extensible options dictionary    
}

job.request reply:
payload: {
   "id"       : "ID"      # On success new, non-zero id returned
   "errmsg"   : "string"  # On failure, an error message is returned
}
```

Given more recent developments, what changes need to be made here? Seems like the "spec" we've been talking about actually could subsume cmdline and resources at the least. Do we need to simplify this to:
```
topic: "job.request" # or, "job.submit" ?
payload: {
  "jobspec" : "string",  # opaque string blob for now
  "cwd"       : "path",
  "environ"   : {  "var": "value", }, # (overrides or complete)
  "options"   : { option: "value" },  # extensible options dictionary    
}
```

Additionally the reply would be extended to allow an optional warning, and include kvs path, and kvs root version instead of ID on success.

The new job module shall also allow registration of flexible callbacks to allow feasibility checks and scheduler registration. 

Above this, I don't have any great design ideas so I'm open to suggestions. For callbacks I was thinking of allowing lua scripts to be sent to job module and run in order (with a "consumed" return code to indicate that the stack should stop processing). This seemed flexible enough to allow for our 3 use cases for these callbacks/filters:
 - modification of job requests to set defaults, etc.
 - job feasibility checks
 - notify scheduler of new jobs and allow sched to optionally reject

Other than that I was going to perform a straightforward replacement of the current job module.

It does occur to me that we might want to keep in mind the following other ideas:
 - is there a better way to make the `job.submit` service more distributed?
 - need some ideas for bulk load/job array submission
 - we need a nice interface to query running, pending, completed jobs (seems like this could be a secondary service that is easy to make distributed)
 - any other good ideas for high throughput enhancements? I was thinking an interesting experiment here would be use a hash of the job req as a storage location in the kvs, allowing any job request to be directly written to kvs without going through the job module. Then you might be able to accomplish some filtering/callbacks in parallel, and then only after these are complete feed job "ids" to singleton entities like the scheduler.

These are probably almost all future work at this point.

**Comment by trws** [comment]
Just a thought, but we could say that the job module is, itself, a distributed service, and registering modules can note whether they need global information or not.  That way, the job.create, job.submit and the first several checks like feasibility could probably happen in the local broker's version of the module before passing the id to a, currently presumably centralized scheduler.  Not sure if there's much more there.  I'm curious about the lua sending option though, what would a user be allowed to do with that lua blob?

**Comment by grondo** [comment]
I wasn't saying there was a lua component to the user blob sent in the request, though some functionality like that may one day be useful (user "preference" function for example, but wouldn't need to be Lua). For now I was thinking filters would be registered as small lua scriptlets that could do one or more of the following
 * modify the job request or jobspec directly (e.g. slurm's job_submit plugin)
 * make a query to resource.* interface as feasibility smoke test
 * make direct rpc's to other services, e.g. `sched` via an api that could reject, consume, or accept jobspec

Your thought about the job module is a good one. In this case we're not going to be able to get the next unique job id though as it is done now (not that I'm saying that is a good method in any way). What were you thinking we'd do with job "ids" in the distributed service case? (Sorry I'm not coming up with any good ideas today)

**Comment by trws** [comment]
When as and if it becomes available, I would go for an asynchronous atomic update somewhere in the kvs to get it, or better yet use a consensus protocol like raft (#370) to get one in a distributed fashion from the get-go.  For right this minute though, we already depend pretty deeply on UUIDs being unique.  It could just be hidden behind a "get_job_id" smokescreen that for now generates UUIDs, switches to a KVS-based scheme when #251 happens, and could later become something else if necessary.

I think all that would actually require is that we treat the ID as either a string or a sized blob rather than an int, but I'm pretty sure we treat them as strings in most places anyway right now, since they're turned into directory IDs, right?

**Comment by garlick** [comment]
Random comment:  maybe a job id should be a UUID plus user-supplied tag?  As long as the tag is unique it could be used as a user-friendly handle for the job (listing output, command input, etc), with abbreviated UUID optionally thrown in to disambiguate?

I'm a fan of avoiding the need for distributed jobid generation if sequential jobids are merely an artifact of past monolithic designs and not a requirement.

**Comment by trws** [comment]
As long as it's unique, I think we're good.  Allowing a user to specify a tag sounds like a good idea to me, it could even be something crazy-simple like `<tag>-<hostname>-<broker-tid>-<timestamp>`, as long as it's unique.  Actually, I've used almost exactly that as the unique primary key on quite a number of database tables to avoid the use of "AUTO_INCREMENT," which tends to result in a global lock and a rather horrific loss of performance...  As a mitigating thing to help users deal with this, something mercurial did well and git did not, we could always provide a per-user job-id as an integer, but I'm not sure we need it.

Anyway, it occurs to me we could just require that all such modules can be run from anywhere if we want to.  As long as there is a way to atomically acquire a reservation globally, and thus find out if someone else has already taken the resources you thought were free, all of the actual logic could be done in a distributed manner, then only contend when actually attempting to run something.  It would be slightly more complicated, and still serialized at the acquire if it's a CAS-like primitive, but much more computationally scalable and relatively simple to write a load-balancer for.  *shrugs* Next steps maybe.

**Comment by grondo** [comment]
Yes, good idea on the uuid, I was suggesting hash of contents because requesting entity could know a priori what the ID would be, but other than there is no benefit and a UUID is just as good (already used in `cmb.exec` service anyway). 

I like the idea of user tag, generated by a filter if not supplied. This will make viewing/searching for jobs much more user friendly. We'd also want a way to at least loosely sort uuids, maybe based on a kvs timestamp?

@trws, you might have to explain your last comment again -- are you talking about development of a distributed scheduler?

**Comment by trws** [comment]
Not exactly.  Making it possible to run several copies of a scheduler 
that are more-or-less stand-alone but act against the same pool of 
resources.  The net result would be a distributed scheduler I suppose, 
but I wasn't thinking of anything quite fancy enough to be worthy of the 
name.  Basically, it's safe to ignore that paragraph as a thought 
experiment.

**Comment by grondo** [comment]
We've got the new job-ingest module so this can be closed.
